### ISSUE TITLE
Question: regarding async-mode

### ISSUE THREAD

**Issue by jeremyschulman** [ncclient/ncclient#245]
Hi Team,

I am using the async-mode to execute a Junos software package add RPC.  This is a long running process, and during the install the Junos response RPC will emit output, which is actually captured within an <output> tag.  I have code written using the async mode AOK; but I am wondering if I can access the "ncclient read buffer" while the RPC is running so that I can get "near real time" progress.  While I know that the buffer string is "incomplete" I could at least use the contents in an intelligent want to provide useful feedback on progress to the User.  

So my code effectively looks like this:

````python
    rpc_rsp = ncc.rpc(rpc_cmd)

    # check the status of the RPC against a countdown timeout ...
    while not rpc_rsp.event.is_set():
        cur_act.details['countdown'] = count_down
        cur_act.updated()
        count_down -= check_after
        if count_down < 0:
            break
        time.sleep(check_after)
````

What I'd like to do is access a property within the `rpc_rsp` object so that I can obtain the read buffer within the while loop.

I hope this question makes sense.  I look forward to your reply and guidance.

Thank you!
-- Jeremy

(ps - I am also back at Juniper)

**Comment by einarnn** [comment]
Jeremy,

tl;dr -- the way ncclient currently receives messages can't let you do what you want without substantial changes.

When receiving messages, the IPC reading code just pulls bytes off of the SSH channel, and looks for either the 1.0 or 1.1 message delimiter, depending on which was selected after the hello. The 1.0 and 1.1 parsing code each do this a little but differently. For 1.0, it just appends to a StringIO buffer until it sees the delimiter, and for 1.1 it builds a list of received chunks. In both cases, the final message is coalesced just before dispatching the message back to the logical session that has knowledge of the outstanding RPCs.

This means that the receiving code has no idea which RPC it is currently reading data for, so it has no way to link the bytes while reading them back to your original RPC. Without major changes, the best that could be notified would be some kind of "I just read N byes from this SSH session", probably to a new "activity listener" or something. And this wouldn't give any _direct_ insight into which RPC the bytes were relevant to or how many more bytes are needed to complete the message. In effect, you'd know that something was happening, but not what.

Also, if one of your goals of moving to async operations is to look at leveraging pipelining, potentially with multiple outstanding operations, then the link from your original request to the bytes currently being received becomes even more tenuous without the inclusion of more bookkeeping infrastructure that would also depend on the underlying server honouring the ordering constraints detailed in [Section 4.5 of RFC 6241](https://tools.ietf.org/html/rfc6241#section-4.5).

One thing I did think about was freezing the buffer currently being appended to and provide a copy back to your code. This freeze & copy would be necessary because the receiving code runs on a separate thread, and so we cannot give direct read access. And, again, unless you parsed this buffer yourself, you'd not know if it was anything to do with your current RPC. Also, the performance hit of doing this for large messages would be substantial.

Given this, I don't think what you are looking for is achievable without fairly major surgery on the code, and rethinking many aspects of how messages are processed. An optional "activity listener", with the limitations I noted above, is probably the only simply added approach.

Cheers,

Einar
